### DISCUSSION TITLE
Added: LDM Notes for July 16, 2018

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#1729] [General]
https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-07-16.md

Agenda:

1. Null-coalescing assignment
   1. User-defined operators
   1. Unconstrained type parameters
   1. Throw expression the right-hand side
1. Nullable await
1. Nullable pointer access
1. Non-nullable reference types feature flag follow-up

**Comment by Logerfo** [comment]
> This is a hole in null-coalescing as well. Both features should be allowed on unconstrained type parameters.

I don't get it. Is this a typo, missing the negation (both features should NOT be allowed), or the null coalescing behavior will change in order for both features be allowed on unconstrained type parameters?

**Comment by gafter** [comment]
This is not a typo. Both features should work for unconstrained type parameters.

**Comment by Joe4evr** [comment]
> or the null coalescing behavior will change in order for both features be allowed on unconstrained type parameters?

Correct, see #1720.

**Comment by bbarry** [comment]
> Let's favor the semantics of assignment. The return type should be the type of `x`.

That seems backwards given that the previous conclusion suggested the semantics should be the same as `(x ?? (x = y))`. Am I reading something wrong?

**Comment by 333fred** [comment]
@bbarry, the type of `(x ?? (x = y))` is also the type of `x`. The type of the left is just `X`, and the type of the right is the type of the `x = y` assignment, which is also just `X`.

**Comment by 333fred** [comment]
More precisely, given `a ?? (a = a0)`, where `A` is the type of `a` and is a nullable type, and `A0` is the underlying nullable type, and that the type of `a0` is `A0`, we have:
`a ?? (a = a0)` (substitute types) -> `A ?? (A = A0)` (simplify assignment) -> `A ?? A` (simplify coalesce) -> `A`
While we accept that there could be a few use cases for the assignment to yield `A0`, we feel that this would break the substitution property of assignment, where you can always substitute the left hand side of an assignment for the entire assignment without changing the containing expression (ie, the same overload would be chosen).

**Comment by paulomorgado** [comment]
> The spec also states that the semantics are equivalent to x = x ?? y, but that should instead be (x ?? (x = y)). **If x is null, the assignment is never performed.**

Shouldn't it be **If x is not null, the assignment is never performed.** instead?

**Comment by svick** [comment]
@paulomorgado You're not the first person to notice that, see https://github.com/dotnet/csharplang/pull/1856.
